### PR TITLE
watchdogd: init at 4.0; add module

### PR DESCRIPTION
**Pull request by vifino**
## Description of changes

https://troglobit.com/projects/watchdogd/

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 1]
```diff
@@ -0,0 +1,128 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+  cfg = config.services.watchdogd;
+
+  mkPluginOpts = plugin: {
+    enable = mkEnableOption "watchdogd plugin ${plugin}";
+    interval = mkOption {
+      type = types.int;
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:9]:
```suggestion
      type = types.ints.unsigned;
```

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+      default = 300;
+      description = ''
+        The poll interval. Defaults to 300 seconds.
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:12]: I don't think the defaults need to be included in the descriptions. They will be printed in the manual anyway.

```suggestion
        Amount of seconds between every poll.
```

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+      '';
+    };
+    logmark = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Log current stats every poll interval.
+        Defaults to disabled.
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:20]:
```suggestion
        Whether to log current stats every poll interval.
```

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    '';
+in {
+  options.services.watchdogd = {
+    enable = mkEnableOption "Advanced system & process supervisor";
+    package = mkOption {
+      default = pkgs.watchdogd;
+      defaultText = "pkgs.watchdogd";
+      type = types.package;
+      description = "watchdogd package to use";
+    };
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:56]:
```suggestion
    package = mkPackageOptionMD pkgs "wtachdogd" { };
```

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+    timeout = mkOption {
+      type = types.int;
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:59]:
```suggestion
      type = types.ints.unsigned;
```

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+      default = 15;
+      description = ''
+        The WDT timeout before reset.
+        Defaults to 15 seconds.
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:63]:
```suggestion
        The WDT timeout before reset.
```

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+      '';
+    };
+    interval = mkOption {
+      type = types.int;
+      default = 5;
+      description = ''
+        The kick interval, i.e. how often watchdogd(8) should reset the WDT timer.
+        Defaults to 5 seconds.
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:71]:
```suggestion
        The kick interval, i.e. how often {manpage}`watchdogd(8)` should reset the WDT timer.
```

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+      '';
+    };
+
+    safeExit = mkOption {
+      type = types.bool;
+      default = true;
+      description = ''
+        With safeExit enabled, the daemon will ask the driver to disable the WDT before exiting.
+        However, some WDT drivers (or HW) may not support this
+        Defaults to true.
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:81]: Some of these abbreviations are a bit hard to follow. WDT is probably fine, as it seems to be a shorthand used for this program (assuming WatchDogTimer?), but I think `HW` is a bit unnecessary.

```suggestion
        With {var}`safeExit` enabled, the daemon will ask the driver to disable the WDT before exiting.
        However, some WDT drivers (or hardware) may not support this
```

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = types.lines;
+      default = "";
+      description = ''
+        Additional config to put in watchdogd.conf.
+        See watchdogd.conf(5) for syntax.
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:90]:
```suggestion
        Additional config to put in {file}`watchdogd.conf`.
        See {manpage}`watchdogd.conf(5)` for syntax.
```

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+
+    # Create the service.
+    systemd.services.watchdogd = {
+      wantedBy = [ "multi-user.target" ];
+      description = "Advanced system & process supervisor";
+      path = [ cfg.package ];
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:106]: Does this program invoke itself assuming it exists in its own `$PATH`? Else I think this line can be removed.

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+      text = ''
+        timeout = ${toString cfg.timeout}
+        interval = ${toString cfg.interval}
+        safe-exit = ${boolToString cfg.safeExit}
+        ${mkPluginConf "filenr"}
+        ${mkPluginConf "loadavg"}
+        ${mkPluginConf "meminfo"}
+        ${extraConfig}
+      '';
+    };
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:126]: I think it would be better if this was written directly to the nix store using something like `writeFile`, instead of doing indirection via `/etc`.

File: pkgs/by-name/wa/watchdogd/package.nix [hunk 12]
```diff
@@ -0,0 +1,24 @@
+{ lib, stdenv, fetchFromGitHub, pkg-config, autoreconfHook, libite, libuev
+, libconfuse }:
+stdenv.mkDerivation rec {
+  pname = "watchdogd";
+  version = "3.5";
+
+  src = fetchFromGitHub {
+    owner = "troglobit";
+    repo = "watchdogd";
+    rev = "${version}";
```
Comment (h7x4) [inline pkgs/by-name/wa/watchdogd/package.nix:10]: Unnecessary string interpolation

```suggestion
    rev = version;
```

File: pkgs/by-name/wa/watchdogd/package.nix [hunk 13]
```diff
@@ -0,0 +1,24 @@
+{ lib, stdenv, fetchFromGitHub, pkg-config, autoreconfHook, libite, libuev
+, libconfuse }:
```
Comment (h7x4) [inline pkgs/by-name/wa/watchdogd/package.nix:2]: This is just a suggestion, but I think it would be nice if these arguments were laid out with newlines in between them. It makes it a lot easier to review with a cleaner diff, if anything were to be added or removed.

```suggestion
{ lib
, stdenv
, fetchFromGitHub
, pkg-config
, autoreconfHook
, libite
, libuev
, libconfuse
}:
```

File: pkgs/by-name/wa/watchdogd/package.nix [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    rev = "${version}";
+    hash = "sha256-DnXIOGZQK16jn20ASwrKIih/61zUuTTgeq7Gv9WLxxU=";
+  };
+
+  nativeBuildInputs = [ pkg-config autoreconfHook ];
+  buildInputs = [ libite libuev libconfuse ];
+
+  meta = with lib; {
+    description = "Advanced system & process supervisor for Linux";
+    homepage = "https://troglobit.com/watchdogd.html";
```
Comment (h7x4) [inline pkgs/by-name/wa/watchdogd/package.nix:19]: Let's add the changelog as well

```suggestion
    homepage = "https://troglobit.com/watchdogd.html";
    changelog = "https://github.com/troglobit/watchdogd/releases/tag/${version}";
```

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    timeout = mkOption {
+      type = types.int;
+      default = 15;
+      description = ''
+        The WDT timeout before reset.
+        Defaults to 15 seconds.
+      '';
+    };
+    interval = mkOption {
+      type = types.int;
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:67]:
```suggestion
      type = types.ints.unsigned;
```

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+    let pcfg = cfg.${plugin};
+    in optionalString pcfg.enable ''
+      ${plugin} {
+        enabled  = true
+        interval = ${toString pcfg.interval}
+        logmark  = ${toString pcfg.logmark}
+        warning  = ${toString pcfg.warning}
+        critical = ${toString pcfg.critical}
+      }
+    '';
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:47]: Not a blocker, but this doesn't seem to be used apart from the `config` block. Would you mind moving it over there for better code locality?

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+      '';
+    };
+    warning = mkOption {
+      type = types.int;
+      description = ''
+        The high watermark level. Alert sent to log.
+      '';
+    };
+    critical = mkOption {
+      type = types.int;
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:30]: I must say I'm not entirely sure what these options are, but I'm assuming they're not supposed to be negative numbers

```suggestion
      type = types.ints.unsigned;
      description = ''
        The high watermark level. Alert sent to log.
      '';
    };
    critical = mkOption {
      type = types.ints.unsigned;
```

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+      '';
+    };
+
+    filenr = mkPluginOpts "filenr";
+    loadavg = mkPluginOpts "loadavg";
+    meminfo = mkPluginOpts "meminfo";
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:100]: This seems a bit unnecessary, considering the `watchdogd` command doesn't seem to have any control unix socket that can be controlled via the cli, or something similar. Are there any cases where a user might want to invoke `watchdogd` from their terminal while it's running in systemd?

File: nixos/modules/services/monitoring/watchdogd.nix (hunk 10 above)
Comment (vifino) [reply nixos/modules/services/monitoring/watchdogd.nix:106]: Do I need to add the package to restartTriggers instead or will that happen automatically?

File: pkgs/by-name/wa/watchdogd/package.nix (hunk 13 above)
Comment (vifino) [reply pkgs/by-name/wa/watchdogd/package.nix:2]: I had it this way but nixpkgs-fmt reformatted it again.

File: nixos/modules/services/monitoring/watchdogd.nix (hunk 18 above)
Comment (vifino) [reply nixos/modules/services/monitoring/watchdogd.nix:100]: It does, `watchdogctl(1)` connects to `/run/watchdogd.sock`.

Comment (h7x4) [reply nixos/modules/services/monitoring/watchdogd.nix:100]: Okay! Feel free to mark as resolved :grin:

File: nixos/modules/services/monitoring/watchdogd.nix (hunk 10 above)
Comment (h7x4) [reply nixos/modules/services/monitoring/watchdogd.nix]: No that will happen automatically, as the systemd unit itself changes

**Comment by vifino** [comment]
Hey @h7x4, mind reviewing this again and getting this merged?

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+      description = ''
+        The high watermark level. Alert sent to log.
+      '';
+    };
+    critical = mkOption {
+      type = types.ints.unsigned;
+      description = ''
+        The critical watermark level. Alert sent to log, followed by reboot or script action.
+      '';
+    };
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:33]: While this might mess up the args of this nice `mkOption` function a bit, I think it would be a lot better user experience for this module if we set some defaults for this. As it stands now, the user would have to set all 6 watermark levels manually, unless I'm reading it wrong. Since the function is never exposed to the user, maybe you could just add two more arguments?

These seem to be the upstream defaults: https://github.com/troglobit/watchdogd/blob/6ae714cd22f24e79270296774e2bfb36570d8209/watchdogd.conf#L92-L165

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+  };
+in {
+  options.services.watchdogd = {
+    enable = mkEnableOption "Advanced system & process supervisor";
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:37]: `mkEnableOption` is a bit weird, because it adds `Whether to enable ${desc}.` as the description. In order for this not to become `Whether to enable Advanced system & process supervisor.`, I think we should change it up. How about this?

```suggestion
    enable = mkEnableOption "watchdogd, an advanced system & process supervisor";
```

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+      default = true;
+      description = ''
+        With {var}`safeExit` enabled, the daemon will ask the driver to disable the WDT before exiting.
+        However, some WDT drivers (or hardware) may not support this.
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = types.lines;
+      default = "";
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:66]: While this is definitely more simple, I think it would be more semantically correct to set it to `nullOr lines`, set the default to `null` and use `optionalString` in `config`. Up to you.

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+      ${extraConfig}
+    '';
+  in mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+
+    # Create the service.
+    systemd.services.watchdogd = {
+      wantedBy = [ "multi-user.target" ];
+      description = "Advanced system & process supervisor";
+      restartTriggers = [ watchdogdConf ];
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:106]: I do think nixos-rebuild will restart the process whenever the unit changes, and the unit changes anytime this file changes due to the command in `ExecStart`. So this is probably superfluous. Probably not a big deal though.

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+      ${mkPluginConf "filenr"}
+      ${mkPluginConf "loadavg"}
+      ${mkPluginConf "meminfo"}
+      ${extraConfig}
+    '';
+  in mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+
+    # Create the service.
+    systemd.services.watchdogd = {
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:103]: Suggestion: I saw these were mentioned in the descriptions, and I think they are relevant enough to the daemon to be put here.

```suggestion
    systemd.services.watchdogd = {
      documentation = [
        "man:watchdogd(8)"
        "man:watchdogd.conf(5)"
      ];
```

File: pkgs/by-name/wa/watchdogd/package.nix [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+, libconfuse
+}:
+stdenv.mkDerivation rec {
+  pname = "watchdogd";
+  version = "3.5";
+
+  src = fetchFromGitHub {
+    owner = "troglobit";
+    repo = "watchdogd";
+    rev = version;
```
Comment (h7x4) [inline pkgs/by-name/wa/watchdogd/package.nix:17]: nit: consider using the finalAttrs pattern. It's supposed to be better in some ways, but I'm not sure about the specifics.

```suggestion
stdenv.mkDerivation (finalAttrs: {
  pname = "watchdogd";
  version = "3.5";

  src = fetchFromGitHub {
    owner = "troglobit";
    repo = "watchdogd";
    rev = finalAttrs.version;
```

**Review by h7x4 — COMMENTED** [review]
Here's a few more comments

File: nixos/modules/services/monitoring/watchdogd.nix (hunk 19 above)
Comment (vifino) [reply nixos/modules/services/monitoring/watchdogd.nix:33]: I skipped this because of the loadavg stuff - it's based on the core count as far as I can tell.
But I added defaults for the two others with the extra two arguments.

File: nixos/modules/services/monitoring/watchdogd.nix (hunk 23 above)
Comment (vifino) [reply nixos/modules/services/monitoring/watchdogd.nix:103]: Huh! What does that do, where does it show up?
Added.

File: pkgs/by-name/wa/watchdogd/package.nix (hunk 24 above)
Comment (vifino) [reply pkgs/by-name/wa/watchdogd/package.nix:17]: finalAttrs helps while overriding as it'll update when you override the version, but I think it's not worth it here, since the only relevant place is in the fetcher, where the hash will change.

File: nixos/modules/services/monitoring/watchdogd.nix (hunk 23 above)
Comment (h7x4) [reply nixos/modules/services/monitoring/watchdogd.nix:108]: It shows up in (among other things) `systemctl status <service>`. An example would be `systemctl status nix-daemon`, which contains both a manpage and link. It's nice to have when you're staring at a dead systemd unit, as a sort of reminder to go read the docs.

File: pkgs/by-name/wa/watchdogd/package.nix (hunk 24 above)
Comment (h7x4) [reply pkgs/by-name/wa/watchdogd/package.nix:17]: Sounds fine by me. It's trivial to fix it up, if it ever becomes an issue again or if someone makes a treewide pr.

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+    warning = mkOption {
+      type = types.ints.unsigned;
+      description = ''
+        The high watermark level. Alert sent to log.
+      '';
+    } // optionalAttrs (defWarn != null) {
+      default = defWarn;
+    };
+    critical = mkOption {
+      type = types.ints.unsigned;
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:31]: On second thought, these two cannot have this type. We even set their defaults to be floats. I'm unsure if they will break without the `.0` suffix on ints.

```suggestion
      type = types.numbers.nonnegative;
      description = ''
        The high watermark level. Alert sent to log.
      '';
    } // optionalAttrs (defWarn != null) {
      default = defWarn;
    };
    critical = mkOption {
      type = types.numbers.nonnegative;
```

Comment (vifino) [reply nixos/modules/services/monitoring/watchdogd.nix]: Good call! I completely overlooked the numbers types.
Thanks!

**Comment by vifino** [comment]
Hey @h7x4, mind looking at this again?

**Comment by h7x4** [comment]
Hello again!

I spent some time tinkering with this module, to make it more in style with [RFC 42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md). That includes putting most of the config into `settings`, and allowing for freeform settings. I've also added a small test, and marked you as the maintainer, hope that's okay with you?

It did come at the cost of setting some defaults that we discussed earlier not to set. But I think it should be fine, considering these are defaults provided by upstream.

Here's the patches: https://gist.github.com/h7x4/215f4d642ca0b0e67b91354f3b4ada4f

**Comment by vifino** [comment]
Hello! :)

I took a look at your patch, it seems very sensible!
Will apply the changes and try to expand the test a bit.

I really appreciate you taking the time to review this!
Your feedback is invaluable!

Thank you!

**Comment by vifino** [comment]
Hey @h7x4, the recursive type exploded in the doc building process because of the infinite recursion. I replaced it with a non-recursive variant.

I changed the test so that it now checks if `watchdogctl` can connect to the daemon and runs the builtin tests, so if they pass, it surely works! :)

File: nixos/modules/services/monitoring/watchdogd.nix [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+          valueType = attrsOf (oneOf [
+            bool
+            int
+            float
+            str
+          ]);
+        in attrsOf (oneOf [
+          valueType
+          (attrsOf valueType)
+        ]);
```
Comment (h7x4) [inline nixos/modules/services/monitoring/watchdogd.nix:54]: This seems wrong. If you're trying to restrict the option to only allow for a single nested layer, I think the `attrsOf` on `valueType` needs to go. Also no need for `oneOf` when there's only two options

```suggestion
        freeformType = let
          valueType = oneOf [
            bool
            int
            float
            str
          ];
        in attrsOf (either valueType (attrsOf valueType));
```

Comment (vifino) [reply nixos/modules/services/monitoring/watchdogd.nix:54]: Yes, that seems more correct, thank you!

**Comment by vifino** [comment]
Alright! This sure went up in quality quite a bit! Anything left to do? :)

**Review by h7x4 — APPROVED** [review]
Not as far as I can see :)
